polaris/experimental/_benchmark_v2.py

```diff
@@ -0,0 +1,247 @@
+        yield "test", self.test
+
+
+class BenchmarkV2Specification(BenchmarkSpecification):
+    _version: ClassVar[Literal[2]] = 2
+
+    dataset: DatasetV2 = Field(exclude=True)
+    split: SplitV2
+    n_classes: dict[ColumnName, int]
+
+    @field_validator("dataset", mode="before")
+    @classmethod
+    def _parse_dataset(
+        cls,
+        v: DatasetV2 | str | dict[str, Any],
+    ) -> DatasetV2:
+        """
+        Allows either passing a Dataset object or the kwargs to create one
+        """
+        match v:
+            case dict():
+                return DatasetV2(**v)
+            case str():
+                return DatasetV2.from_json(v)
+            case DatasetV2():
+                return v
+
+    @model_validator(mode="after")
+    def _validate_n_classes(self) -> Self:
+        """
+        The number of classes for each of the target columns.
+        """
+        columns = set(self.n_classes.keys())
+        if not columns.issubset(self.target_cols):
+            raise InvalidBenchmarkError("Not all specified class numbers were found in the target columns.")
+
+        return self
+
+    @model_validator(mode="after")
+    def _validate_split_in_dataset(self) -> Self:
+        """
+        Verifies that:
+          - All indices are valid given the dataset
+        """
+        dataset_length = len(self.dataset)
+        if self.split.max_index >= dataset_length:
+            raise InvalidBenchmarkError("The predefined split contains invalid indices")
+
+        return self
+
+    @computed_field
+    @property
+    def n_train_datapoints(self) -> int:
+        """The size of the train set."""
+        return self.split.n_train_datapoints
+
+    @computed_field
+    @property
+    def n_test_sets(self) -> int:
+        """The number of test sets"""
+        return self.split.n_test_sets
+
+    @computed_field
+    @property
+    def n_test_datapoints(self) -> dict[str, int]:
+        """The size of (each of) the test set(s)."""
+        return self.split.n_test_datapoints
+
+    @computed_field
+    @property
+    def test_set_sizes(self) -> dict[str, int]:
+        return {label: index_set.datapoints for label, index_set in self.split.test_items()}
+
+    @computed_field
+    @property
+    def test_set_labels(self) -> list[str]:
+        return list(label for label, _ in self.split.test_items())
+
+    def _get_test_sets(
+        self, hide_targets=True, featurization_fn: Callable | None = None
+    ) -> dict[str, Subset]:
+        """
+        Construct the test set(s), given the split in the benchmark specification. Used
+        internally to construct the test set for client use and evaluation.
+        """
+        # TODO: We need a subset class that can handle very large index sets without copying or materializing all of them
+        return {
+            label: self._get_subset(index_set.indices, hide_targets, featurization_fn)
+            for label, index_set in self.split.test_items()
+        }
```

Comment:
**Comment:** So this works out of the box with the subset class? That's pretty cool! 😎

Reply:
It does. I do have concerns about how it would handle truly large index sets. I tested the rest with a test/training split of ~1 million indices each. I need to test the Subset class to see how well it handles it.